res/controllers/Reloop-Mixage.scripts.js

```diff
@@ -0,0 +1,784 @@
+        "fx_on": 0x16,
+        "fx_sel": 0x15,
+        "scratch_active": 0x12,
+        "scroll_active": 0x11,
+        "vu_meter": 0x1E,
+    }
+};
+
+// Maps mixxx controls to a function that toggles their LEDs
+Mixage.connectionMap = {
+    "cue_indicator": {"function": function(v, g, c) { Mixage.toggleLED(v, g, c); }},
+    "cue_default": {"function": function(v, g, c) { Mixage.toggleLED(v, g, c); }},
+    "play_indicator": {"function": function(v, g, c) { Mixage.toggleLED(v, g, c); Mixage.toggleLED(v, g, "load_indicator"); }},
+    "pfl": {"function": function(v, g, c) { Mixage.toggleLED(v, g, c); }},
+    "loop_enabled": {"function": function(_v, g) { Mixage.toggleReloopLED(g); }},
+    "loop_in": {"function": function(v, g) { if (v === 1) { Mixage.toggleLoopLED(g); } }},
+    "loop_out": {"function": function(v, g) { if (v === 1) { Mixage.toggleLoopLED(g); } }},
+    "sync_enabled": {"function": function(v, g, c) { Mixage.toggleLED(v, g, c); }},
+    "eject": {"function": function(_v, g) { Mixage.eject(g); }},
+};
+
+// ----- Internal variables functions -----
+
+// Set or remove functions to call when the state of a mixxx control changes
+Mixage.connectControlsToFunctions = function(group, remove) {
+    for (const control in Mixage.connectionMap) {
+        if (remove !== undefined) {
+            Mixage.connectionMap[control][group].disconnect();
+        } else {
+            Mixage.connectionMap[control][group] = engine.makeConnection(group, control, Mixage.connectionMap[control].function);
+        }
+    }
+};
+
+Mixage.init = function(_id, _debugging) {
+
+    // all button LEDs off
+    for (let i = 0; i < 255; i++) {
+        midi.sendShortMsg(0x90, i, 0);
+    }
+
+    // find controls and make engine connections for each channel in Mixage.channels
+    // A predefined list with channels is used instead of a for loop to prevent engine connections to be overwritten
+    Mixage.channels.forEach(function(channel) {
+        const deck = script.deckFromGroup(channel);
+        Mixage.connectControlsToFunctions(channel);
+
+        // set soft takeovers for effectslot amount
+        for (let effectSlot = 1; effectSlot <= Mixage.numEffectSlots; effectSlot++) {
+            const groupString = "[EffectRack1_EffectUnit"+ deck +"_Effect" + effectSlot + "]";
+            engine.softTakeover(groupString, "meta", true);
+        }
+
+        for (let effectUnit = 1; effectUnit <= Mixage.numEffectUnits; effectUnit++) {
+            // make connections for the fx on LEDs
+            const fxGroup = "group_"+channel+"_enable";
+            Mixage.fxOnConnection.push(engine.makeConnection("[EffectRack1_EffectUnit"+effectUnit+"]", fxGroup, function() { Mixage.toggleFxLED(channel); }));
+
+            // set soft takeovers for effectunit meta
+            engine.softTakeover("[EffectRack1_EffectUnit"+effectUnit+"]", "super1", true);
+            engine.setValue("[EffectRack1_EffectUnit"+effectUnit+"]", "show_focus", 1);
```

Comment:
some renaming, deduplication, and 2.4 supports template literals I think.
```suggestion
            const fxControl = `group_${channel}_enable`;
            const fxGroup = "[EffectRack1_EffectUnit${effectUnit}]"
            Mixage.fxOnConnection.push(engine.makeConnection(fxGroup, fxControl, function() { Mixage.toggleFxLED(channel); }));

            // set soft takeovers for effectunit meta
            engine.softTakeover(fxGroup, "super1", true);
            engine.setValue(fxGroup, "show_focus", 1);
```

Reply:
Coming from a python background I was wondering if template literals where available in JS as it would greatly improve code readability. I wrote this code initially in es5 (mixxx 2.3) which as far as I could find didn't support this feature. With the move to es6 with 2.4 this should indeed now be possible. I will change the string concatination to template literals where applicable.

Reply:
I replaced all concatenations with string literals which should improve code readability

Reply:
I'm resolving this. Feel free to comment if more thins pop up